Migrate from pipenv to poetry (#84)
niteoweb/pyramid-realworld-example-app: readme.md

Pyramid serves one of the [RealWorld.io frontends](https://github.com/gothinkster/realworld#frontends) (Elm) on root, so it is easier to understand how things fit together. The frontend is interchangeable, you can use any
RealWorld.io frontend.

# Getting started

To setup your development environment, there are some dependencies you need to install on your system.
You can either install them manually, or use [Nix](https://nixos.org/nix) and [direnv](https://direnv.net/)
to install them automatically (except docker which you need to install manually).

So, first ensure that you have [Docker](https://www.docker.com) installed on your system. Then,

### Option 1: Install manually

- Exact version of Python mentioned in `./pyproject.toml` (recommend using [pyenv](https://github.com/pyenv/pyenv) for that).
- [poetry](https://github.com/python-poetry/poetry)

> - [Node(js)](https://nodejs.org) & [jq](https://github.com/stedolan/jq) (to run postman tests)

Reviewer: Specify versions for these pls.

Reviewer: Or tell people to check `shell.nix` for exact version requirements.

Reviewer: Even better. Say that they need to install Python 3.7.X and Node 11.X, etc. And that if they have problems, then check exact versions in shell.nix.

Reviewer: Even even better: Create a small shell script that bumps versions declared in README.md automatically every time we run `make lint` or something.

Author: > Create a small shell script that bumps versions mentioned in README.md automatically every time we run make lint or something.

Feels overkill because poetry will tell us that it needs the correct version of Python if not installed.

^ Tested by writing the Python version as `~3.9` in `pyproject.toml`.
```
[EnvCommandError]
Command ['/Users/sayan/Documents/GitHub/pyramid-realworld-example-app/.venv/bin/pip', 'install', '-e', '/Users/sayan/Documents/GitHub/pyramid-realworld-example-app'] errored with the following return code 1, and output:
Obtaining file:///Users/sayan/Documents/GitHub/pyramid-realworld-example-app
ERROR: Package 'conduit' requires a different Python: 3.8.1 not in '>=3.9,<3.10'
```
Also, I'd like poetry to be the source of truth since `shell.nix` is optional here. poetry will raise error if the Python version in `shell.nix` is out of sync.

> Even better. Say that they need to install Python 3.7.X and Node 11.X, etc. And that if they have problems, then check exact versions in shell.nix.

This will also work I guess.

Author: > Feels overkill 

I mean we need to maintain a specific parsable format of Python version in the readme. Also we'll be depending upon the `bash`/`sed` version in different platforms.
Also, we might need to write test cases for different platforms. Else, some untested code will be overwriting our readme file.

Reviewer: > I mean we need to maintain a specific parsable format of Python version in the readme.

Yes, that is correct. But since we control the README.md, this shouldn't be too hard.

> Also we'll be depending upon the bash/sed version in different platforms. Also, we might need to write test cases for different platforms. Else, some untested code will be overwriting our readme file.

Not true, we use nix-shell in development and in CI, so we can use bash/sed that is pinned by `shell.nix`.

> Also, I'd like poetry to be the source of truth since shell.nix is optional here. poetry will raise error if the Python version in shell.nix is out of sync.

Yes, this is true for Python. But Poetry does not know anything about which Nodejs and jq versions are needed. So these two need to come from `shell.nix`.

Author: I don't think we need to specify nodejs and jq version for our use case. We don't specify that even in `shell.nix`. For Python version, I'd go with the 2nd approach i.e. ` Say that they need to install Python 3.7.X and Node 11.X, etc`. We don't upgrade Python that often. When we do, we need to update a bunch of files anyway. Readme can be one small part of that.

### Option 2: Install using Nix

- Install the [Nix package manager](https://nixos.org/nix).

> - Install and configure [Direnv](https://direnv.net/) (you need to `direnv allow` the project root).

Reviewer: ```suggestion
- Install and configure [Direnv](https://direnv.net/)
- Run `direnv allow` in the project root.
```

# Getting started

> You need to have [Docker](https://www.docker.com), [Nix](https://nixos.org/nix) and [direnv](https://direnv.net/) installed and configured on your system.

Reviewer: Nix and direnv should be optional. The Readme should first give instructions how to install dependencies on Ubuntu and MacOS (using `apt-get` and `brew`) and finally tell the user about nix, as a B-plan if they have problems.

### Option 1: Install manually

- [python 3.8](https://www.python.org) (recommend using [pyenv](https://github.com/pyenv/pyenv) for installing Python).
- [poetry](https://github.com/python-poetry/poetry)
- [node](https://nodejs.org) & [jq](https://github.com/stedolan/jq) (to run postman tests)

> If you have any dependency related issue, check out `./shell.nix` for the specific version of dependencies.

Reviewer: Good enough!
